File: ormmakurippu/ios/CallTrackeriOS/CallTrackeriOS/ViewModels/CallViewModel.swift
import Foundation
import Combine
import FirebaseAuth
import UIKit

@MainActor
class CallViewModel: ObservableObject {
    @Published var calls: [Call] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    // Dashboard Stats
    @Published var totalCount = 0
    @Published var newCount = 0
    @Published var sentCount = 0
    
    private let service = FirebaseService.shared
    private var cancellables = Set<AnyCancellable>()
    
    var userId: String? {
        Auth.auth().currentUser?.uid
    }
    
    func fetchCalls() {
        guard let userId = userId else { return }
        isLoading = true
        
        Task {
            do {
                let fetchedCalls = try await service.fetchCalls(for: userId)
                self.calls = fetchedCalls
                self.calculateStats()
                self.isLoading = false
            } catch {
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
        }
    }
    
    private func calculateStats() {
        totalCount = calls.count
        newCount = calls.filter { $0.status == "NEW" }.count
        sentCount = calls.filter { $0.status == "SENT" }.count
    }
    
    func addCall(name: String, phone: String, assignee: String, notes: String, attachment: UIImage?) {
        guard let userId = userId else { return }
        isLoading = true
        
        Task {
            do {
                var attachmentUrl: String? = nil
                if let image = attachment {
                    attachmentUrl = try await service.uploadAttachment(image: image, userId: userId)
                }
                
                let call = Call(
                    name: name,
                    phone: phone,
                    date: Date(),
                    assignee: assignee,
                    status: "NEW",
                    notes: notes,
                    attachmentUrls: attachmentUrl != nil ? [attachmentUrl!] : nil,
                    createdAt: Date(),
                    createdByUid: userId
                )
                
                try service.addCall(call)
                await fetchCalls() // Refresh
                self.isLoading = false
            } catch {
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
        }
    }
    
    func updateStatus(call: Call, newStatus: String) {
        guard let id = call.id else { return }
        Task {
            do {
                try await service.updateStatus(callId: id, newStatus: newStatus)
                // Optimistic update
                if let index = calls.firstIndex(where: { $0.id == id }) {
                    calls[index].status = newStatus
                    calculateStats()
                }
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }
    
    func deleteCall(call: Call) {
        guard let id = call.id else { return }
        Task {
            do {
                try await service.deleteCall(callId: id)
                 if let index = calls.firstIndex(where: { $0.id == id }) {
                    calls.remove(at: index)
                    calculateStats()
                }
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }
}

File: android/CallTrackerAndroid/app/src/main/java/com/example/calltracker/viewmodel/CallViewModel.kt
package com.example.calltracker.viewmodel

import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.calltracker.data.model.CallEntity
import com.example.calltracker.data.repository.CallRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import com.example.calltracker.data.service.AnalysisResult
import java.util.Date
import java.util.UUID
import javax.inject.Inject

@HiltViewModel
class CallViewModel @Inject constructor(
    private val repository: CallRepository
) : ViewModel() {

    // All calls from repository
    val allCalls = repository.getAllCalls()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _searchQuery = MutableStateFlow("")
    val searchQuery = _searchQuery.asStateFlow()

    private val _statusFilter = MutableStateFlow<String?>(null)
    val statusFilter = _statusFilter.asStateFlow()

    // Filtered calls
    val filteredCalls = combine(allCalls, searchQuery, statusFilter) { calls, query, status ->
        calls.filter { call ->
            val matchesQuery = call.name.contains(query, ignoreCase = true) ||
                             call.phone.contains(query, ignoreCase = true)
            val matchesStatus = status == null || call.status == status
            matchesQuery && matchesStatus
        }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    // Dashboard stats
    val stats = allCalls.combine(MutableStateFlow(0)) { calls, _ ->
        val total = calls.size
        val newCount = calls.count { it.status == "NEW" }
        val sentCount = calls.count { it.status == "SENT" }
        Triple(total, newCount, sentCount)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), Triple(0, 0, 0))

    fun refresh() {
        viewModelScope.launch {
            repository.syncCalls()
        }
    }

    fun setSearchQuery(query: String) {
        _searchQuery.value = query
    }

    fun setStatusFilter(status: String?) {
        _statusFilter.value = status
    }

    fun addCall(name: String, phone: String, assignee: String, notes: String, attachmentUri: Uri?) {
        viewModelScope.launch {
            val call = CallEntity(
                id = UUID.randomUUID().toString(),
                name = name,
                phone = phone,
                date = Date(),
                assignee = assignee,
                status = "NEW",
                notes = notes,
                attachmentUrls = emptyList(),
                createdAt = Date(),
                createdByUid = "", // Repo handles this
                needsSync = true
            )
            repository.saveCall(call, attachmentUri)
        }
    }

    fun updateCallStatus(call: CallEntity, newStatus: String) {
        viewModelScope.launch {
            repository.saveCall(call.copy(status = newStatus))
        }
    }
    
    fun deleteCall(callId: String) {
        viewModelScope.launch {
            repository.deleteCall(callId)
        }
    }

    suspend fun getCallById(id: String): CallEntity? {
        return repository.getCallById(id)
    }

    // AI Features
    private val _analysisResult = MutableStateFlow<AnalysisResult?>(null)
    val analysisResult = _analysisResult.asStateFlow()

    fun analyzeNotes(notes: String) {
        viewModelScope.launch {
            _analysisResult.value = repository.analyzeNotes(notes)
        }
    }

    fun clearAnalysis() {
        _analysisResult.value = null
    }
}
